tests/ui/unsafe-fields.rs

```diff
@@ -0,0 +1,100 @@
+//@ compile-flags: --crate-type=lib
+#![allow(incomplete_features)]
+#![feature(unsafe_fields)]
+
+struct WithUnsafeField {
+    unsafe unsafe_field: u32,
+    safe_field: u32,
+}
+
+enum A {
+    WithUnsafeField { unsafe unsafe_field: u32, safe_field: u32 },
+}
+
+fn f(a: A) {
+    let A::WithUnsafeField { unsafe_field, safe_field } = a;
+    //~^ ERROR
+}
+
+struct WithInvalidUnsafeField {
+    unsafe unsafe_noncopy_field: Vec<u32>, //~ ERROR
+}
+
+struct WithManuallyDropUnsafeField {
+    unsafe unsafe_noncopy_field: std::mem::ManuallyDrop<Vec<u32>>,
+}
+
+union WithUnsafeFieldUnion {
+    unsafe unsafe_field: u32,
+    safe_field: u32,
+}
+
+impl WithUnsafeField {
```

Comment:
Can you add a test tracking the behavior of this:
```rust
&raw const self.unsafe_field
```
I *think* (but am not completely certain) it's safe, but we should probably require `unsafe` in the initial implementation.

Reply:
I *think* I would prefer this to be `unsafe` -- of course, to do something with the result of this that could possibly cause problems requires some other form of unsafe somewhere else (I think...), but making this `unsafe` is IMO nice because it improves locality-of-safety:
```
let ptr = &raw const self.unsafe_field;

// lots of code

unsafe { do_something_with_ptr }
```
I think it'd be nice to have some sign somewhere that `ptr` actually comes from an unsafe field.